proxy/coco/src/peer/subroutines.rs

```diff
@@ -0,0 +1,299 @@
+                self.state.clone(),
+                self.store.clone(),
+                self.input_sender.clone(),
+            )),
+            Command::Control(control_command) => match control_command {
+                ControlCommand::Respond(respond_command) => {
+                    SpawnAbortable::new(control_respond(respond_command))
+                },
+            },
+            Command::SyncPeer(peer_id) => {
+                SpawnAbortable::new(sync(self.state.clone(), peer_id, self.input_sender.clone()))
+            },
+            Command::StartSyncTimeout(sync_period) => {
+                SpawnAbortable::new(start_sync_timeout(sync_period, self.input_sender.clone()))
+            },
+            Command::Request(RequestCommand::Query(urn)) => {
+                SpawnAbortable::new(query(urn, self.state.clone(), self.input_sender.clone()))
+            },
+            Command::Request(RequestCommand::Clone(url)) => {
+                SpawnAbortable::new(clone(url, self.state.clone(), self.input_sender.clone()))
+            },
+        }
+    }
+}
+
+impl Drop for Subroutines {
+    fn drop(&mut self) {
+        for task in self.pending_tasks.iter_mut() {
+            task.abort()
+        }
+    }
+}
+
+impl Future for Subroutines {
+    type Output = Result<(), spawn_abortable::Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        // Drain the task queue.
+        loop {
+            match self.pending_tasks.poll_next_unpin(cx) {
+                Poll::Ready(Some(Err(e))) => {
+                    log::warn!("error in spawned subroutine task: {:?}", e);
+                    return Poll::Ready(Err(e));
+                },
+                Poll::Ready(Some(Ok(()))) => continue,
+                // Either pending, or FuturesUnordered thinks it's done, but
+                // we'll enqueue new tasks below
+                Poll::Ready(None) | Poll::Pending => break,
+            }
+        }
+
+        // Drain all pending inputs, feed them to the [`RunState`] and execute the returned
+        // commands as async tasks.
+        loop {
+            match self.inputs.poll_next_unpin(cx) {
+                Poll::Ready(Some(input)) => {
+                    log::debug!("handling subroutine input: {:?}", input);
+
+                    if let Some(event) = Event::maybe_from(&input) {
+                        // Ignore if there are no subscribers.
+                        self.subscriber.send(event).ok();
+                    }
+
+                    for cmd in self.run_state.transition(input) {
+                        let task = self.spawn_command(cmd);
+
+                        self.pending_tasks.push(task);
+                    }
+                },
+                Poll::Ready(None) => return Poll::Ready(Ok(())),
+                Poll::Pending => {
+                    if !self.pending_tasks.is_empty() {
+                        cx.waker().wake_by_ref()
```

Comment:
I believe this will make the executor spin and repeatedly poll this future. Let’s assume we have one task in `self.pending_tasks` that is blocked by IO and will return `Poll::Pending` for the next couple of seconds. Let’s also assume that there are now inputs arriving on `self.inputs`. If you follow the code in this method if will always break out of the first loop that calls `self.pending_tasks.poll_next_unpin(cx)` and land here. Since we are calling the waker, the task that runs this future will schedule the `poll()` here to be called again. This brings us back to the begining. And since nothing has changed (we’re still waiting on IO from the other task) we will be stuck in a loop.

The code wants to address the following issue

>  Once self.inputs returns Pending and we did enqueue new tasks, we should summon the waker before returning Pending.

This is handled by `FuturesUnordered`. We are passing the context to `FuturesUnordered::poll_next_unpin()`. If one of the tasks in `self.pending_tasks` finishes, `FuturesUnordered` will call the waker which ensures that this future is polled again.

Reply:
I don't think your reasoning is correct:

a. There is no "make the executor spin" -- calling `wake_by_ref` puts us on the back of the poll queue, nothing more. Iow, its semantics is "yield".
b. Morally, we **want** to recurse here, because our `FuturesUnordered` was most likely empty before we entered the second `loop` -- ie. it returned `Ready` from being polled. However, since we **know** that we enqueued new tasks, we must ensure that we are polled again soonish, in order to poll the inputs queue.
c. If there are no new inputs **nor** have enqueued new tasks, we don't wake -- because there's nothing new we can expect when polled again. So, we only return `Pending`.

Reply:
Oh I see, we want to distinguish between `Pending` and `Ready` as the result of polling the `FuturesUnordered` -- it it was `Pending`, it wasn't empty before.

Reply:
The second part of my analysis was wrong. If `pending_tasks.poll_next_unpin()` returns `Poll::Ready(None)` we will need to ensure that `Subroutine::poll` will be called again if we push an item to `pending_tasks`. This is because `FuturesUnordered` will not wake up any task when the pushed future finishes. I guess this was the point that you made, @kim.

> Oh I see, we want to distinguish between `Pending` and `Ready` as the result of polling the `FuturesUnordered` -- it it was `Pending`, it wasn't empty before.

Indeed. It is exactly this case that makes the executor spin. The executor will poll every future on its queue and these polls are inexpensive and don’t block. So even if we are at the back and there are a lot of other futures queued we will call `Subroutine::poll` again pretty quickly. And if nothing has changed (in contrast to `yield`) we will put the future on the queue again.

Reply:
eeacd3c